browser/brave_ads/search_result_ad/search_result_ad_handler.cc

```diff
@@ -0,0 +1,293 @@
+        const auto* it = std::find(kSearchResultAdAttributes.begin(),
+                                   kSearchResultAdAttributes.end(),
+                                   base::StringPiece(ad_property->name));
+
+        if (it == kSearchResultAdAttributes.end()) {
+          LOG(ERROR) << "Wrong search ad attribute specified: "
+                     << ad_property->name;
+          return SearchResultAdsList();
+        }
+        found_attributes.insert(*it);
+
+        if (!SetSearchAdProperty(ad_property, search_result_ad.get())) {
+          LOG(ERROR) << "Cannot read search ad attribute value: "
+                     << ad_property->name;
+          return SearchResultAdsList();
+        }
+      }
+
+      // Not all of attributes were specified.
+      if (found_attributes.size() != kSearchResultAdAttributes.size()) {
+        std::vector<base::StringPiece> absent_attributes;
+        std::set_difference(kSearchResultAdAttributes.begin(),
+                            kSearchResultAdAttributes.end(),
+                            found_attributes.begin(), found_attributes.end(),
+                            std::back_inserter(absent_attributes));
+
+        LOG(ERROR) << "Some of search ad attributes were not specified: "
+                   << base::JoinString(absent_attributes, ", ");
+
+        return SearchResultAdsList();
+      }
+
+      search_result_ads_list.push_back(std::move(search_result_ad));
+    }
+
+    // Creatives has been parsed.
+    break;
+  }
+
+  return search_result_ads_list;
+}
+
+SearchResultAdsList ParseMetadataEntities(blink::mojom::WebPagePtr web_page) {
+  for (const auto& entity : web_page->entities) {
+    if (entity->type != kSearchAdLinkDataType) {
+      continue;
+    }
+
+    absl::optional<SearchResultAdsList> search_result_ads_list =
+        ParseEntityProperties(entity);
+
+    if (search_result_ads_list) {
+      return std::move(*search_result_ads_list);
+    }
+  }
+
+  return SearchResultAdsList();
+}
+
+}  // namespace
+
+SearchResultAdHandler::SearchResultAdHandler() = default;
+
+SearchResultAdHandler::~SearchResultAdHandler() = default;
+
+bool SearchResultAdHandler::IsAllowedBraveSearchHost(const GURL& url) const {
+  if (!url.is_valid() || !url.SchemeIs(url::kHttpsScheme)) {
+    return false;
+  }
+  const std::string host = url.host();
+  return base::Contains(kVettedBraveSearchHosts, host);
+}
+
+void SearchResultAdHandler::RetrieveSearchResultAd(
+    content::RenderFrameHost* render_frame_host,
+    RetrieveSearchResultAdCallback callback) {
+  DCHECK(render_frame_host);
+  DCHECK(IsAllowedBraveSearchHost(render_frame_host->GetLastCommittedURL()));
+  DCHECK(!document_metadata_.is_bound());
+
+  render_frame_host->GetRemoteInterfaces()->GetInterface(
+      document_metadata_.BindNewPipeAndPassReceiver());
+  DCHECK(document_metadata_.is_bound());
+
+  document_metadata_->GetEntities(
+      base::BindOnce(&SearchResultAdHandler::OnRetrieveSearchResultAdEntities,
```

Comment:
weakptr is fine here, but since you're calling a method on a mojo interface that is owned by this class it's not strictly necessary. base::Unretained is safe here